src/cdomain/value/cdomains/valueDomain.ml

```diff
@@ -353,6 +354,22 @@ struct
       IntDomain.Size.is_cast_injective ~from_type:t1 ~to_type:t2 && bitsSizeOf t2 >= bitsSizeOf t1
     | _ -> false
 
+  (* is a cast t1 to t2 invertible, i.e., content-preserving for the given value of v? *)
+  let is_dynamically_safe_cast t2 t1 v =
+    if is_statically_safe_cast t2 t1 then
+      true
+    else
+      match t2, t1, v with
+      | (TInt (ik2,_) | TEnum ({ekind=ik2; _},_)) , (TInt (ik1,_) | TEnum ({ekind=ik1; _},_)), Int v ->
+        let cinterval = ID.cast_to ik2 v in
+        let cl, cu = ID.minimal cinterval, ID.maximal cinterval in
+        let l, u = ID.minimal v, ID.maximal v in
+        (match cl,cu,l,u with
+         | Some cl, Some cu, Some l, Some u when Z.leq cl l && Z.leq u cu -> true
```

Comment:
Doesn't `ID.cast_to` sometimes go to top when the target type doesn't fit the current values?
Because then the condition would trivially hold even though all values didn't fit.

Reply:
> I think this should be safe, and it would even be safe if we chose the bounds of `ik2` for `cl` and `cu` as far as I can see.

But then I don't see the need to check against the casted range of `v` as opposed to `ik2` bounds. Checking whether `(l,u)` is included in the casted range (if it's not `ik2` bounds), then this check is just unnecessarily stricter, if the bounds are already sound.

For example, interval domain overflow check does just that. So we could skip the redundant `ID.cast_to` computation here.

Reply:
That is of course true! I didn't know that `Size` was accessible from outside the `IntDomain` module.